posts/2020-11-27-typing-your-actors.adoc

```diff
@@ -0,0 +1,389 @@
+      val user2      = Room.User(UUID.randomUUID(), "user 2", false, "", user2Probe.ref)
+
+      behaviorTestKit.run(
+        RoomManager
+          .ConnectToRoom(WSMessage(MessageType.Join, roomId, user1.id, user1.name), user1Probe.ref)
+      )
+      behaviorTestKit.run(
+        RoomManager
+          .ConnectToRoom(WSMessage(MessageType.Join, roomId, user2.id, user2.name), user2Probe.ref)
+      )
+
+      val childInbox = behaviorTestKit.childInbox[Room.Command](roomId.toString)
+      childInbox.expectMessage(Room.Join(user1))
+      childInbox.expectMessage(Room.Join(user2))
+    }
+
+    "handle an IncomeWSMessage that generates an outcome" in {
+      val roomId    = UUID.randomUUID()
+      val roomProbe = testKit.createTestProbe[Room.Command]()
+      val managerRef = testKit.spawn(
+        RoomManager.receiveBehaviour(RoomManagerData(Map(roomId -> roomProbe.ref)))
+      )
+      val userId = UUID.randomUUID()
+
+      managerRef ! RoomManager.IncomeWSMessage(WSMessage(MessageType.Vote, roomId, userId, "5"))
+      managerRef ! RoomManager.IncomeWSMessage(
+        WSMessage(MessageType.EditIssue, roomId, userId, "issue name")
+      )
+      managerRef ! RoomManager.IncomeWSMessage(WSMessage(MessageType.Show, roomId, userId, ""))
+      managerRef ! RoomManager.IncomeWSMessage(WSMessage(MessageType.Clear, roomId, userId, ""))
+
+      roomProbe.expectMessage(Room.Vote(userId, "5"))
+      roomProbe.expectMessage(Room.EditIssue(userId, "issue name"))
+      roomProbe.expectMessage(Room.ShowVotes(userId))
+      roomProbe.expectMessage(Room.ClearVotes(userId))
+    }
+----
+
+RoomSpec:
+
+[source,scala]
+----
+"Room Actor" should {
+    "update current issue and broadcast it" in {
+      val issue               = "Issue test 1"
+      val (user, userProbe)   = createUser(UUID.randomUUID(), "user1", false, "")
+      val (user2, user2Probe) = createUser(UUID.randomUUID(), "user2", false, "")
+      val dataProbe           = testKit.createTestProbe[Room.Response]()
+      val actingUserId        = UUID.randomUUID()
+      val (roomId, roomRef) = createRoom(
+        UUID.randomUUID(),
+        RoomData.empty.copy(users = List(user, user2))
+      )
+
+      val expectedMessage = WSMessage(MessageType.EditIssue, roomId, actingUserId, issue)
+      val expectedData = Room.DataStatus(data =
+        RoomData(
+          users = List(user, user2),
+          currentIssue = issue,
+          issueLastEditBy = Option(actingUserId)
+        )
+      )
+
+      roomRef ! Room.EditIssue(actingUserId, issue)
+
+      roomRef ! Room.GetData(dataProbe.ref)
+
+      userProbe.expectMsg(expectedMessage)
+      user2Probe.expectMsg(expectedMessage)
+
+      dataProbe.expectMessage(expectedData)
+    }
+  }
+----
+
+Takeaways:
+
+* Since actors are `Behavior` functions overriding internal functions for testing is not suggested like in https://doc.akka.io/docs/akka/current/testing.html#overriding-behavior[classical].
```

Comment:
```suggestion
* Since actors are `Behavior` functions, overriding internal functions for testing is not suggested like in https://doc.akka.io/docs/akka/current/testing.html#overriding-behavior[classical].
```